notebooks/day2/metaprogramming.jl

```diff
@@ -0,0 +1,1103 @@
+above; it passes the tuple `(expr1, expr2, ...)` as one argument to the macro:
+
+```julia
+@name (expr1, expr2, ...)
+```
+"""
+
+# ╔═╡ a1e86478-5e66-11eb-2ef9-6d460e707013
+md"""
+### Building an advanced macro
+
+Here is a simplified definition of Julia's `@assert` macro, which checks if expression is true:
+
+(`... ? ... : ...` is the ternary operator seen in `01-basics.jl` )
+"""
+
+# ╔═╡ e61381aa-5e66-11eb-3347-5d26b61e6c17
+macro assert(ex)
+	return :( $ex ? nothing : throw(AssertionError($(string(ex)))) )
+end
+
+# ╔═╡ a43490de-5e67-11eb-253e-f7455ade3ffe
+md"""This macro can be used like this:"""
+
+# ╔═╡ c79ad04c-5e67-11eb-1fd9-bf7ff40f0862
+md"""
+In place of the written syntax, the macro call is expanded at parse time to its returned result. This is equivalent to writing:"""
+
+# ╔═╡ f34e9020-5e67-11eb-1987-e3e6a0969705
+1 == 1.0 ? nothing : throw(AssertionError("1 == 1.0"))
+
+# ╔═╡ f81955fe-5e67-11eb-11b9-5f9fd0096a09
+1 == 0 ? nothing : throw(AssertionError("1 == 0"))
+
+# ╔═╡ fb772ef4-5e67-11eb-218e-330ba087bcc8
+md"""
+That is, in the first call, the expression `:(1 == 1.0)` is spliced into the test condition slot,
+while the value of `string(:(1 == 1.0))` is spliced into the assertion message slot. The entire
+expression, thus constructed, is placed into the syntax tree where the `@assert` macro call occurs.
+Then at execution time, if the test expression evaluates to true, then `nothing` is returned,
+whereas if the test is false, an error is raised indicating the asserted expression that was false.
+Notice that it would not be possible to write this as a function, since only the *value* of the
+condition is available and it would be impossible to display the expression that computed it in
+the error message.
+
+The actual definition of `@assert` in Julia Base is more complicated. It allows the
+user to optionally specify their own error message, instead of just printing the failed expression.
+Just like in functions with a variable number of arguments, this is specified with an ellipses
+following the last argument:
+"""
+
+# ╔═╡ 2df2baa8-5e68-11eb-2f28-9968b7ecbcd5
+macro assert(ex, msgs...)
+	msg_body = isempty(msgs) ? ex : msgs[1]
+	msg = string(msg_body)
+	return :($ex ? nothing : throw(AssertionError($msg)))
+end
+
+# ╔═╡ b0dfd0dc-5e67-11eb-19cf-5970a2e80bfa
+@assert 1 == 1.0
+
+# ╔═╡ c42d2d6a-5e67-11eb-1564-2dcb0e91fb19
+@assert 1 == 0
+
+# ╔═╡ 376fe8e4-5e68-11eb-1d7d-c9b1945b133f
+md"""
+Now `@assert` has two modes of operation, depending upon the number of arguments it receives!
+If there's only one argument, the tuple of expressions captured by `msgs` will be empty and it
+will behave the same as the simpler definition above. But now if the user specifies a second argument,
+it is printed in the message body instead of the failing expression. You can inspect the result
+of a macro expansion with the aptly named `@macroexpand`
```

Comment:
There is something strange with this sentence.

Reply:
@DaanVanHauwermeiren this PR can be merged after fixing this, I'm too stupid to construct a proper sentence from this

Reply:
solved in f85d321